### DISCUSSION TITLE
Contribute to GNU Emacs, suitable to replace ispell.el in use

### DISCUSSION THREAD

**Discussion by rrthomas** [minad/jinx#200] [General]
`jinx.el` says "This file is part of GNU Emacs", but I don't find the file in Emacs git, so I assume this is not (yet) true. I would however very much like it to be!

I am the upstream maintainer of Enchant, and the author of Emacs's built-in Enchant support via `ispell.el`. As the upstream maintainer of Enchant, I would like to remove the `enchant` program on which `ispell.el` depends, and as the author of the Enchant support in `ispell.el`, I would like to see this old and crufty module edged towards retirement, or at least be a spell-checker of last resort for those who need or are used to its particular functionality.

Until today I had not used Jinx, but I found it easy to set up and it seems to work pretty well.

So I have a few questions:

- Would you @minad be amenable to contributing Jinx to GNU Emacs? I am guessing that the copyright notice in `jinx.el` is not quite right either, but assuming you're the only significant author, it would be easy to make a copyright assignment. If you're not, then I'd be happy to help audit the code. As a rough guide, contributions of less than 15 lines of code don't need a copyright assignment.)
- What's missing, if anything, for Jinx to be a reasonable replacement for `ispell.el`? I discount `ispell.el`'s special support for parsing certain modes' buffers, such as mail or LaTeX, as that in my view is the wrong way around these days—the modes should be coming to the spell-checker, not the other way around. After some cursory use of Jinx myself, the one thing I don't see quite how to do is a formal "whole-buffer" check, à la `ispell`. I guess one can use `jinx-correct` with `M-n` and `M-p`, but that doesn't quite put one in a formal workflow. Perhaps that doesn't matter and I'm just old-fashioned? But I can imagine that being off-putting to users in a way that the lack of `ispell.el`'s more esoteric functionality wouldn't be.

**Comment by minad** [comment]
(Converted to a discussion)

> jinx.el says "This file is part of GNU Emacs", but I don't find the file in Emacs git, so I assume this is not (yet) true. I would however very much like it to be!

As far as I understand, all ELPA packages can be legally considered part of Emacs, that's why the header is written as is. But I am also open to contribute my packages to be part of upstream itself.

There have been a few brief discussions of my packages on emacs-devel, but no concrete steps towards an inclusion. As far as I understand, the policy is to rather improve ispell.el or flyspell.el, which is certainly possible, if one accepts the addition of even more cruft, given that existing modes of operation shouldn't be affected. In contrast, Jinx is a new start and just a thin layer on top of Enchant.

> As the upstream maintainer of Enchant, I would like to remove the enchant program on which ispell.el depends, and as the author of the Enchant support in ispell.el

I am personally not using the enchant program, but I'd strongly recommend against removing it. The ispell protocol is still the de-facto standard.

> I would like to see this old and crufty module edged towards retirement, or at least be a spell-checker of last resort for those who need or are used to its particular functionality.

In principle I agree that ispell.el is full of cruft and should be retired. But from what I've heard such a deprecation is not going to happen, and it also doesn't have to happen. Users who prefer the ispell way can continue to use it as is, while users who prefer the Jinx-style can use that instead. The packages are complementary and not mutually exclusive.

> Until today I had not used Jinx, but I found it easy to set up and it seems to work pretty well.

Thanks.

> Would you @minad be amenable to contributing Jinx to GNU Emacs? I am guessing that the copyright notice in jinx.el is not quite right either, but assuming you're the only significant author, it would be easy to make a copyright assignment.

The copyright assignment is correct. The package is assigned to the FSF by me. See also my other ELPA packages, all of them are assigned to the FSF. Jinx is not my first package, and also not my most popular one.

> What's missing, if anything, for Jinx to be a reasonable replacement for ispell.el? I discount ispell.el's special support for parsing certain modes' buffers, such as mail or LaTeX, as that in my view is the wrong way around these days—the modes should be coming to the spell-checker, not the other way around. 

Jinx relies on font locking, regexps and custom predicate functions to restrict the checked text. I think the existing mechanisms are mostly sufficient, but I've heard that TeX support could be better. I am not fond of pulling in any legacy code or mechanisms, but improvements within the current design are certainly welcome.

> After some cursory use of Jinx myself, the one thing I don't see quite how to do is a formal "whole-buffer" check, à la ispell. I guess one can use jinx-correct with M-n and M-p, but that doesn't quite put one in a formal workflow. 

You can run `jinx-correct-all` and correct all misspellings in the whole buffer one by one.

**Comment by minad** [comment]
I've seen your issue https://github.com/AbiWord/enchant/issues/393. I hadn't been aware that the enchant program was only created with Emacs/ispell.el support in mind. I thought that the enchant program had always been meant as replacement for the ispell/aspell/hunspell command line programs, also in other scenarios. This certainly changes the equation a little bit, and I understand why you would want to deprecate the command line program (EDIT: If I understand correctly, you only want to remove the ispell protocol via the -a command line option?). Nevertheless my recommendation would still be to keep it, also given that it is useful for other integrations and debugging. But of course, it depends on the maintenance burden.

Regarding ispell.el and making Jinx an adequate replacement for ispell.el/flyspell.el - I wonder which specific functionality do you consider missing? By design, I think Jinx cannot be an exact replacement  and doesn't aim to be one. However it should be possible to find equivalent workflows for everything done in ispell/flyspell.

**Comment by rrthomas** [comment]
Thanks very much for your detailed comments!

Indeed, the command-line program lacks the interactive abilities of aspell/ispell etc. and I've never seen any demand for that functionality (e.g. a disappointed user who thought they would be able to do this with `enchant`). It would certainly be possible to deprecate it instead, and of course it could be forked by another project, not least because it's implemented entirely with the public API. I could even spin it out myself as a "free gift" to future maintainers.

I don't have specific examples of missing functionality in Jinx, and I agree also that it would not be necessary to add like-for-like functionality to consider it a suitable replacement for `ispell.el`; I'm just concerned that most current use-cases be covered (as you say, "equivalent workflows"), and I guessed there might be gaps. `jinx-correct-all` was the only one I thought I'd found, but you pointed it out.

**Reply by shipmints** [reply]
I like having the command-line enchant tool. It helps me know if libenchant is working, and what version. strace and dtrace are helpful to verify shared library loading.

**Reply by rrthomas** [reply]
There would still be a command-line tool that loads the libraries, like `enchant-lsmod` now.

**Reply by shipmints** [reply]
enchant-lsmod doesn't actually check words so hard to know if it's truly working. But I hear you on at least the dynamic library utility.

**Reply by rrthomas** [reply]
`enchant -l` checks spellings, and that's easy to keep too.

**Reply by shipmints** [reply]
Perhaps I misunderstood, then. I do test with enchant-2 so I guess that's the same one.

**Comment by rrthomas** [comment]
So, it seems like there's nothing formally to do here. Jinx is already FSF-assigned and in ELPA; the maintainers could add it to Emacs if they wanted to. I agree that realistically, `ispell.el` isn't going to be dropped any time soon (or even, as I would like to see, deprecated in favour of Jinx); the Emacs maintainers are (sadly in my view) resistant to pushing packages out of core into ELPA (why they don't do this with large standalone packages such as AuCTeX and org-mode baffles me!).

I will continue to see how Jinx works for me, and file more precise issues as and when, but I think this discussion can be closed. Based on feedback to my issue about dropping `enchant -a`, I think I will target that change for Enchant 3, the notional release that will "go full gintrospection", as being a suitable point to implement a breaking change/simplification.

**Comment by rrthomas** [comment]
I found one other thing: `LocalWords` support, which I understand is possible from https://github.com/minad/jinx/issues/12 , would be highly desirable to have built in for backwards and cross-compatibility with `ispell.el`. I can see this being a pain point otherwise, as imposing a switching cost.

**Reply by minad** [reply]
Yes, I've intentionally not added `LocalWords` support, given that one can use the more general file local variables instead. However I agree that this leads to a switching cost, so I may reconsider.

EDIT: I've looked into the LocalWords writing/reading code from the wiki again. I cannot adopt this code in this form, given that it depends on ispell.el, which I don't want to load in Jinx. Besides that I am not completely opposed to adding this. I think it is not nice to have a separate feature for loading file local variables, given that Emacs comes with a dedicated mechanism. I tried to avoid this kind of legacy cruft in Jinx.

**Reply by rrthomas** [reply]
It seems reasonable to load ispell code if one is using ispell functionality; if one doesn't use it, it won't be needed. Or is the issue that the `require` would have to be moved to the top level? In which case, supplying this functionality in a separate module `jinx-ispell` or somesuch would seem reasonable.

**Reply by minad** [reply]
Yes, I think we would have to load ispell, something which I want to avoid, since Jinx is free standing. A separate feature `jinx-ispell` would work, but I have no interest in maintaining this. I would however accept a reimplementation in Jinx itself.

**Reply by rrthomas** [reply]
I think it would be a shame to duplicate the implementation of a legacy feature (it would however be quite straightforward simply to copy and paste the code from `ispell.el`, and I don't think that using `ispell.el` code makes Jinx any less standalone, since `ispell.el` is a core part of Emacs.

**Reply by minad** [reply]
Yes, I am just distinguishing loading ispell.el and using parts from it.

**Reply by rrthomas** [reply]
So you'd be happy to copy and paste code from `ispell.el`? As I say, I would prefer not to do that, but if that's what it takes to get jinx supporting `LocalWords` out of the box and enable users to transition easily from ispell to jinx, then I'm game!

**Reply by astoff** [reply]
The `LocalWords` mechanism doesn't look technically very good to me. For example, it doesn't work with a `.dir-locals.el` file.

Ideally Emacs would define some generic variables for local words, dictionary selection, and also a mechanism for major modes to have a say on what parts of the buffer should be checked (instead of it being the spellchecking package's duty to support all major modes). Then different spellchecking packages could make use of that API.

**Reply by minad** [reply]
> The LocalWords mechanism doesn't look technically very good to me.

Yes, it is not great, since it reinvents the local variables wheel and is not sufficient because we cannot specify the language/dictionary. So @rrthomas, "happy" isn't the right word. ;)

> Ideally Emacs would define some generic variables for local words, dictionary selection, ...

Yes, I had thought about this too. Adding generic variables like `local-words` or `local-language` would be my favorite solution.

> ...and also a mechanism for major modes to have a say on what parts of the buffer should be checked (instead of it being the spellchecking package's duty to support all major modes).

I think the face-based mechanism is quite good already. Additional APIs impose a cost on major mode authors.

**Reply by rrthomas** [reply]
> The `LocalWords` mechanism doesn't look technically very good to me. For example, it doesn't work with a `.dir-locals.el` file.

I agree, but I think that improving per-file wordlists should be considered separately from making it seamless to move from ispell to jinx.
